std-bits/table/src/main/java/org/enso/table/operations/Distinct.java

```diff
@@ -50,4 +53,42 @@ public static BitSet buildDistinctRowsMask(
 
     return mask;
   }
+
+  public static BitSet buildDuplicatesRowsMask(
+      int tableSize,
+      Column[] keyColumns,
+      TextFoldingStrategy textFoldingStrategy,
+      ProblemAggregator problemAggregator) {
+    ColumnAggregatedProblemAggregator groupingProblemAggregator =
+        new ColumnAggregatedProblemAggregator(problemAggregator);
+    Context context = Context.getCurrent();
+    var mask = new BitSet();
+    if (keyColumns.length != 0) {
+      Map<MultiValueKeyBase, Integer> visitedRows = new HashMap<>();
+      int size = keyColumns[0].getSize();
+      Storage<?>[] storage =
+          Arrays.stream(keyColumns).map(Column::getStorage).toArray(Storage[]::new);
+      List<TextFoldingStrategy> strategies = ConstantList.make(textFoldingStrategy, storage.length);
+      for (int i = 0; i < size; i++) {
+        UnorderedMultiValueKey key = new UnorderedMultiValueKey(storage, i, strategies);
+        key.checkAndReportFloatingEquality(
+            groupingProblemAggregator, columnIx -> keyColumns[columnIx].getName());
+
+        var keyIndex = visitedRows.get(key);
+        if (keyIndex == null) {
+          visitedRows.put(key, i);
+        } else {
+          mask.set(i);
+          mask.set(keyIndex);
+        }
+
+        context.safepoint();
+      }
+    } else {
```

Comment:
Can we make the no-op at the Table level?
This would cause the whole thing to be duplicated.

Reply:
It's actually unreachable already from the table level. We throw if there are no columns selected.